### PR TITLE
[WIP] Binance staking

### PR DESCRIPTION
**Pull request by catenocrypt**
Staking info from APIs.  Mocked tests.
Open: APR computation

### CODE REVIEW

**Comment by catenocrypt** [comment]
For testing:

```
make go-build
make start-api-mock
curl http://localhost:3347/mock/binance-api/staking/validators?status=bonded
curl http://localhost:3347/mock/binance-api/v1/staking/chains/0/delegators/bnb1jeu6gscugy6l2wyatxthkh2hmer4hzevgcmf0q/delegations

curl http://localhost:8437/v2/binance/staking/validators
curl -d '[{"coin": 714}]' http://localhost:8437/v2/staking/list
curl http://localhost:8437/v2/binance/staking/delegations/bnb1jeu6gscugy6l2wyatxthkh2hmer4hzevgcmf0q?Authorization=Bearer
curl -d '[{"coin": 714, "address": "bnb1jeu6gscugy6l2wyatxthkh2hmer4hzevgcmf0q"}]' http://localhost:8437/v2/staking/delegations
```

File: pkg/blockatlas/client.go [hunk 1]
```diff
@@ -109,7 +110,9 @@ func (r *Request) Execute(method string, url string, body io.Reader, result inte
 	if err != nil {
 		return errors.E(err, errors.TypePlatformRequest)
 	}
-
+	if res.StatusCode != 200 {
```
Comment (EnoRage) [inline pkg/blockatlas/client.go:113]: do not inject any dependency here

File: pkg/blockatlas/client.go [hunk 2]
```diff
@@ -5,14 +5,15 @@ import (
...
 	"fmt"
-	"go.elastic.co/apm/module/apmhttp"
 	"io"
 	"io/ioutil"
 	"net/http"
 	"net/url"
 	"strings"
 	"time"
 
+	"go.elastic.co/apm/module/apmhttp"
```
Comment (EnoRage) [inline pkg/blockatlas/client.go:15]: do not inject any dependency here

**Comment by catenocrypt** [comment]
Put back to Draft, need to add unit tests

File: platform/binance/client.go [hunk 3]
```diff
@@ -98,3 +99,55 @@ func (c Client) FetchTokens() (Tokens, error) {
...
+	}
+	c.Cache.Set("validators", *result, cache.DefaultExpiration)
+	return *result, nil
+}
+
+func (c *Client) GetDelegations(chainID string, address string) (delegations []Delegation, err error) {
+	path := fmt.Sprintf("/api/v1/staking/chains/%s/delegators/%s/delegations", chainID, address)
+	resp, err := req.Get(c.url + path)
+	if err != nil {
+		return []Delegation{}, err
```
Comment (EnoRage) [inline platform/binance/client.go:129]: return nil, err

File: platform/binance/client.go [hunk 4]
```diff
@@ -98,3 +99,55 @@ func (c Client) FetchTokens() (Tokens, error) {
...
+	}
+	var result DelegationsResponse
+	if err := resp.ToJSON(&result); err != nil {
+		logger.Error("URL: " + resp.Request().URL.String())
+		logger.Error("Status code: " + resp.Response().Status)
+		return []Delegation{}, err
```
Comment (EnoRage) [inline platform/binance/client.go:135]: return nil, err

File: platform/binance/client.go [hunk 5]
```diff
@@ -98,3 +99,55 @@ func (c Client) FetchTokens() (Tokens, error) {
...
+	}
+	return result.Delegations, nil
+}
+
+func (c *Client) GetUnbondingDelegations(chainID string, address string) (delegations []UnbondingDelegation, err error) {
+	path := fmt.Sprintf("/api/v1/staking/chains/%s/delegators/%s/ubds", chainID, address)
+	resp, err := req.Get(c.url + path)
+	if err != nil {
+		return []UnbondingDelegation{}, err
```
Comment (EnoRage) [inline platform/binance/client.go:144]: return nil, err

File: platform/binance/client.go [hunk 6]
```diff
@@ -98,3 +99,55 @@ func (c Client) FetchTokens() (Tokens, error) {
...
+	}
+	var result UnbondingDelegationResponse
+	if err := resp.ToJSON(&result); err != nil {
+		logger.Error("URL: " + resp.Request().URL.String())
+		logger.Error("Status code: " + resp.Response().Status)
+		return []UnbondingDelegation{}, err
```
Comment (EnoRage) [inline platform/binance/client.go:150]: return nil, err

File: platform/binance/stake.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+	"github.com/trustwallet/blockatlas/pkg/logger"
+	"github.com/trustwallet/blockatlas/services/assets"
+)
+
+const (
+	chainID       = "bsc"
+	lockTime      = 86400 // in seconds
+	minimumAmount = "0.0001"
+	dummyAPR      = 0.2 // TODO
```
Comment (EnoRage) [inline platform/binance/stake.go:16]: what is that?

File: platform/binance/stake.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	return results, nil
+}
+
+func (p *Platform) UndelegatedBalance(address string) (string, error) {
+	accountMeta, err := p.client.FetchAccountMeta(address)
+	if err != nil {
+		return "0", err
+	}
+	for _, coin := range accountMeta.Balances {
+		if coin.Symbol == "BNB" {
```
Comment (EnoRage) [inline platform/binance/stake.go:84]: coin.Binance().Symbol

File: platform/binance/stake.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+	results = append(results, NormalizeDelegations(delegations, validators)...)
+	results = append(results, NormalizeUnbondingDelegations(unbondingDelegations, validators)...)
+
+	return results, nil
+}
+
+func (p *Platform) UndelegatedBalance(address string) (string, error) {
+	accountMeta, err := p.client.FetchAccountMeta(address)
+	if err != nil {
+		return "0", err
```
Comment (EnoRage) [inline platform/binance/stake.go:81]: Why "0"?

**Review by EnoRage — COMMENTED** [review]
Good job, just small suggestions.

Are you sure that we need to implement it for BNB nor for BSC? We are using Ethereum package for BSC. 

If staking is at BSC - we need to move it there then

File: platform/binance/stake.go (hunk 9 above)
Comment (catenocrypt) [reply platform/binance/stake.go:81]: UndelegatedBalance interface method returns amount as string

https://github.com/trustwallet/blockatlas/blob/master/pkg/blockatlas/platform.go#L47

**Comment by catenocrypt** [comment]
> Are you sure that we need to implement it for BNB nor for BSC? We are using Ethereum package for BSC.
> If staking is at BSC - we need to move it there then

Staking happens on both BC&BSC, but rewards are recorded on BC.
I followed @hewig guidence.  Also here:
"all the staking related operations are recorded in BC"
https://docs.binance.org/guides/concepts/bc-staking.html

File: pkg/blockatlas/client.go (hunk 2 above)
Comment (catenocrypt) [reply pkg/blockatlas/client.go:15]: Sorry, this was like that.  I just added an extra error check to this generic file, not changing it's structure.

**Comment by EnoRage** [comment]
> > Are you sure that we need to implement it for BNB nor for BSC? We are using Ethereum package for BSC.
> > If staking is at BSC - we need to move it there then
> 
> Staking happens on both BC&BSC, but rewards are recorded on BC.
> I followed @hewig guidence. Also here:
> "all the staking related operations are recorded in BC"
> https://docs.binance.org/guides/concepts/bc-staking.html

If so, maybe we need to move it to the BSC. There is a lot of logic at /ethereum package, so maybe not. Let's discuss it.
@hewigovens @vikmeup what do you think?

**Comment by hewigovens** [comment]
In our case, it's always happens on binance chain (like Cosmos delegation), user can delegate/undelegate BNB (binance chain) to some validator to earn rewards

**Comment by EnoRage** [comment]
So, 

> In our case, it's always happens on binance chain (like Cosmos delegation), user can delegate/undelegate BNB (binance chain) to some validator to earn rewards

So, we need to move stake api to the ETH, also make it available only for BSC 

It is more logical to have staking at BSC not BNB if it is about BSC

**Comment by hewigovens** [comment]
yes better to move to standalone `bsc` package? I don't think Ethereum staking will be like this

**Comment by hewigovens** [comment]
Is this pr ready to test?

**Comment by catenocrypt** [comment]
@hewigovens , yes, it is ready for testing (no longer draft).  Only open issue is the APR computation -- it is missing still.  Also, there was the question about moving it to separate package.

**Comment by hewigovens** [comment]
APR is quite low and there's no api to get it yet

File: platform/binance/client.go [hunk 10]
```diff
@@ -98,3 +99,55 @@ func (c Client) FetchTokens() (Tokens, error) {
 	c.Cache.Set("tokens", *result, cache.DefaultExpiration)
 	return *result, nil
 }
+
+func (c *Client) GetValidators() (validators ValidatorsResponse, err error) {
+	cachedResult, ok := c.Cache.Get("validators")
```
Comment (vikmeup) [inline platform/binance/client.go:104]: Use `c.GetWithCache(&result, "staking/pool", nil, time.Minute*20)`

Example: https://github.com/trustwallet/blockatlas/blob/e90b7df9a3bb607b9791da377e313fe97a82d3f5/platform/kava/client.go#L36

File: platform/binance/client.go [hunk 11]
```diff
@@ -98,3 +99,55 @@ func (c Client) FetchTokens() (Tokens, error) {
...
+	query := url.Values{
+		"status": {"bonded"},
+	}
+	result := new(ValidatorsResponse)
+	resp, err := req.Get(c.url+"/api/v1/staking/validators", query)
+	if err != nil {
+		return *result, err
+	}
+	if err := resp.ToJSON(&result); err != nil {
+		logger.Error("URL: " + resp.Request().URL.String())
```
Comment (vikmeup) [inline platform/binance/client.go:117]: remove logging or combine into one

File: platform/binance/client.go [hunk 12]
```diff
@@ -98,3 +99,55 @@ func (c Client) FetchTokens() (Tokens, error) {
...
+	}
+	query := url.Values{
+		"status": {"bonded"},
+	}
+	result := new(ValidatorsResponse)
+	resp, err := req.Get(c.url+"/api/v1/staking/validators", query)
+	if err != nil {
+		return *result, err
+	}
+	if err := resp.ToJSON(&result); err != nil {
```
Comment (vikmeup) [inline platform/binance/client.go:116]: use `https://github.com/trustwallet/blockatlas/blob/e90b7df9a3bb607b9791da377e313fe97a82d3f5/platform/kava/client.go#L41`

you don't need to have custom json parsing

File: platform/binance/client.go [hunk 13]
```diff
@@ -98,3 +99,55 @@ func (c Client) FetchTokens() (Tokens, error) {
...
+	c.Cache.Set("validators", *result, cache.DefaultExpiration)
+	return *result, nil
+}
+
+func (c *Client) GetDelegations(chainID string, address string) (delegations []Delegation, err error) {
+	path := fmt.Sprintf("/api/v1/staking/chains/%s/delegators/%s/delegations", chainID, address)
+	resp, err := req.Get(c.url + path)
+	if err != nil {
+		return nil, err
+	}
```
Comment (vikmeup) [inline platform/binance/client.go:130]: same as above suggestion

File: platform/binance/model.go [hunk 14]
```diff
@@ -100,6 +101,43 @@ type (
...
 	}
+
+	Validator struct {
+		Status  int    `json:"status"`
+		Address string `json:"validator"`
+		ValName string `json:"valName"`
+	}
+
+	ValidatorsResponse struct {
+		Total      int
```
Comment (vikmeup) [inline platform/binance/model.go:112]: `Total` unused

File: platform/binance/model.go [hunk 15]
```diff
@@ -100,6 +101,43 @@ type (
...
+
+	Delegation struct {
+		Value            float64 `json:"amount"`
+		DelegatorAddress string  `json:"delegator"`
+		ValidatorAddress string  `json:"validator"`
+	}
+
+	DelegationsResponse struct {
+		Delegations []Delegation `json:"delegations"`
+		Total       int          `json:"total"`
```
Comment (vikmeup) [inline platform/binance/model.go:124]: `Total` unused

File: platform/binance/model.go [hunk 16]
```diff
@@ -100,6 +101,43 @@ type (
...
+		CompleteHeight int     `json:"completeHeight"`
+		Delegator      string  `json:"delegator"`
+		Denom          string  `json:"denom"`
+		InitialBalance float64 `json:"initialBalance"`
+		Validator      string  `json:"validator"`
+	}
+
+	UnbondingDelegationResponse struct {
+		UnbondingDelegations []UnbondingDelegation `json:"unbondingDelegations"`
+		Total                int                   `json:"total"`
```
Comment (vikmeup) [inline platform/binance/model.go:139]: review unused values and delete them

File: platform/binance/stake.go [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+		Details: blockatlas.StakingDetails{
+			Reward:        blockatlas.StakingReward{Annual: reward},
+			MinimumAmount: minimumAmount,
+			LockTime:      lockTime,
+			Type:          blockatlas.DelegationTypeDelegate,
+		},
+	}
+}
+
+func getUnknownValidator(address string) blockatlas.StakeValidator {
```
Comment (vikmeup) [inline platform/binance/stake.go:147]: did you consider reuse this? lots of duplication for `Decommissioned` validator

File: platform/binance/stake_test.go [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+)
+
+func TestPlatform_GetDetails(t *testing.T) {
+	server := httptest.NewServer(createMockedAPI())
+	defer server.Close()
+	p := Init(server.URL)
+	result := p.GetDetails()
```
Comment (vikmeup) [inline platform/binance/stake_test.go:15]: add spacing between setup and do asserts to easily read code

**Comment by catenocrypt** [comment]
Becoming stale, closing

**Pull request closed without merging**